File: backends/spring-boot/src/main/java/com/hellotimepro/springboot/repository/RefreshTokenRepository.java
package com.hellotimepro.springboot.repository;

import com.hellotimepro.springboot.domain.RefreshTokenEntity;
import jakarta.persistence.LockModeType;
import java.time.OffsetDateTime;
import java.util.Optional;
import java.util.UUID;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface RefreshTokenRepository extends JpaRepository<RefreshTokenEntity, UUID> {
  Optional<RefreshTokenEntity> findByTokenHash(String tokenHash);

  @Lock(LockModeType.PESSIMISTIC_WRITE)
  @Query("select r from RefreshTokenEntity r where r.tokenHash = :tokenHash")
  Optional<RefreshTokenEntity> findByTokenHashForUpdate(@Param("tokenHash") String tokenHash);

  @Modifying
  @Query("update RefreshTokenEntity r set r.revokedAt = :now where r.familyId = :familyId and r.revokedAt is null")
  int revokeFamily(@Param("familyId") UUID familyId, @Param("now") OffsetDateTime now);

  @Modifying
  @Query("update RefreshTokenEntity r set r.revokedAt = :now where r.userId = :userId and r.revokedAt is null")
  int revokeUser(@Param("userId") UUID userId, @Param("now") OffsetDateTime now);
}

File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/repository/RefreshTokenRepository.kt
package com.hellotimepro.ktor.repository

import com.hellotimepro.ktor.db.RefreshTokens
import com.hellotimepro.ktor.db.isSqliteDialect
import com.hellotimepro.ktor.domain.RefreshToken
import org.jetbrains.exposed.sql.*
import java.time.OffsetDateTime
import java.util.UUID

class RefreshTokenRepository {
    fun findByTokenHash(hash: String): RefreshToken? =
        RefreshTokens.select { RefreshTokens.tokenHash eq hash }.firstOrNull()?.toRefreshToken()

    /** Postgres 用行锁保证同一个 refresh token 只能被一个请求完成轮转。 */
    fun findByTokenHashForUpdate(hash: String): RefreshToken? {
        val q = RefreshTokens.select { RefreshTokens.tokenHash eq hash }
        if (!isSqliteDialect) q.forUpdate()
        return q.firstOrNull()?.toRefreshToken()
    }

    fun insert(token: RefreshToken) {
        RefreshTokens.insert {
            it[id] = token.id
            it[userId] = token.userId
            it[tokenHash] = token.tokenHash
            it[familyId] = token.familyId
            it[expiresAt] = token.expiresAt
            it[createdAt] = token.createdAt
            // revokedAt 不写：DB 默认 NULL，避免对自定义列绑定 null。
        }
    }

    fun markRevoked(id: UUID, now: OffsetDateTime) {
        RefreshTokens.update({ RefreshTokens.id eq id }) { it[revokedAt] = now }
    }

    fun revokeFamily(familyId: UUID, now: OffsetDateTime) {
        RefreshTokens.update({ (RefreshTokens.familyId eq familyId) and RefreshTokens.revokedAt.isNull() }) {
            it[revokedAt] = now
        }
    }

    fun revokeUser(userId: UUID, now: OffsetDateTime) {
        RefreshTokens.update({ (RefreshTokens.userId eq userId) and RefreshTokens.revokedAt.isNull() }) {
            it[revokedAt] = now
        }
    }
}
